tests/net/mod.rs

```diff
@@ -0,0 +1,782 @@
+                }
+
+                dest.push_back(NetworkMessage::new(
+                    sender.clone(),
+                    tmsg.message.clone(),
+                    to.clone(),
+                ));
+            }
+            // Broadcast messages get expanded into multiple direct messages.
+            messaging::Target::All => for to in nodes.keys() {
+                if *to == sender {
+                    continue;
+                }
+
+                if !faulty {
+                    message_count = message_count.saturating_add(1);
+                }
+
+                dest.push_back(NetworkMessage::new(
+                    sender.clone(),
+                    tmsg.message.clone(),
+                    to.clone(),
+                ));
+            },
+        }
+    }
+
+    // Collect all outputs (not required for network operation) as a convenience for the user.
+    nodes
+        .get_mut(&sender)
+        .expect("Trying to process a step with non-existing node ID")
+        .outputs
+        .extend(step.output.iter().cloned());
+
+    message_count
+}
+
+/// New network node construction information.
+///
+/// Helper structure passed to node constructors when constructing virtual networks.
+#[derive(Debug)]
+pub struct NewNodeInfo<D>
+where
+    D: DistAlgorithm,
+{
+    /// The node ID for the new node.
+    pub id: D::NodeUid,
+    /// Network info struct, containing keys and other information.
+    pub netinfo: NetworkInfo<D::NodeUid>,
+    /// Whether or not the node is marked faulty.
+    pub faulty: bool,
+}
+
+/// Virtual network builder.
+///
+/// The `NetBuilder` is used to create `VirtualNet` instances and offers convenient methods to
+/// configure the construction process.
+///
+/// Note that, in addition to the constructor `new`, either `using` or `using_step` must be called,
+/// otherwise the construction will fail and panic.
+pub struct NetBuilder<D, I>
+where
+    D: DistAlgorithm,
+{
+    /// Iterator used to create node ids.
+    node_ids: I,
+    /// Number of faulty nodes in the network.
+    num_faulty: usize,
+    /// Constructor function.
+    cons: Option<Box<Fn(NewNodeInfo<D>) -> (D, Step<D>)>>,
+    /// Network adversary.
+    adversary: Option<Box<dyn Adversary<D>>>,
+    /// Trace-enabling flag. `None` means use environment.
+    trace: Option<bool>,
+    /// Optional crank limit.
+    crank_limit: Option<usize>,
+    /// Optional message limit.
+    message_limit: Option<usize>,
+}
+
+impl<D, I> NetBuilder<D, I>
+where
+    D: DistAlgorithm,
+    D::Message: Clone,
+    D::Output: Clone,
+    I: IntoIterator<Item = D::NodeUid>,
+{
+    /// Construct a new network builder.
+    ///
+    /// `node_ids` must be an iterator of the node ids to use for the new node. In many test cases,
+    /// a `Range(usize)` is a convenient value:
+    ///
+    /// ```rust,ignore
+    /// let builder = NetBuilder::new(0..10) // ...
+    /// ```
+    #[inline]
```

Comment:
The compiler automatically inlines within a crate, so they are technically not necessary. I add them on short functions habitually, in case they are ever called cross-crate.

Also, `#[inline]` is just allowing functions to be inline (as opposed to `#[inline(always)]` or `#[inline(never)]`. Putting them there has, in this case, no effect if not called from the outside, if I am not mistaken.

I am really hoping the default changes at some point...